### PR TITLE
feat: add debug plugin collections

### LINKED ISSUE #1250: feat: add built-in debug plugins for generation, validation, and sampling
## Summary
Add comprehensive built-in debug plugins to Mellea for tracing and troubleshooting the full lifecycle of generation, validation, and sampling pipelines.

## What's New
Three categories of debug plugins in `mellea.plugins.builtin_debug`:

### Generation Pipeline (`generation.py`)
- `log_generation_pre_call`: Logs model ID, generation ID, prompt preview, and repair feedback
- `log_generation_post_call`: Logs response preview, latency (ms), token usage, and model

### Validation Pipeline (`validation.py`)
- `log_validation_pre_check`: Logs requirement count and target type
- `log_validation_post_check`: Logs per-requirement pass/fail status with reasons and scores

### Sampling Pipeline (`sampling.py`)
- `log_sampling_loop_start`: Logs strategy name, loop budget, requirement count
- `log_sampling_iteration`: Logs iteration number, pass/fail counts, per-requirement status
- `log_sampling_repair`: Logs repair trigger point, repair type, failed validations
- `log_sampling_loop_end`: Logs success/failure, iterations used, and statistics

## Documentation
- New how-to guide: `docs/docs/how-to/debug-with-plugins.md` covering all plugin categories, usage patterns, and 6 common debugging scenarios
- 7 runnable example scripts in `docs/examples/plugins/` demonstrating each plugin category and combinations

## Examples
- `builtin_generation_tracing.py` — Basic generation pipeline tracing
- `builtin_validation_tracing.py` — Requirement validation with detailed per-requirement results
- `builtin_validation_failures.py` — Real validation failures demonstration
- `builtin_validation_strict.py` — Strict requirements testing
- `builtin_sampling_diagnostics.py` — Sampling strategy diagnostics with repair events
- `builtin_full_pipeline_tracing.py` — End-to-end generation + sampling visibility
- `builtin_complete_diagnostics.py` — All three plugin categories together

## Quality Checks
- ✅ All code passes `ruff format` and `ruff check`
- ✅ All code passes `mypy` type checking
- ✅ All 7 examples execute successfully
- ✅ Documentation passes markdownlint validation

## Usage
```python
from mellea.plugins.builtin_debug.generation import (
    log_generation_pre_call,
    log_generation_post_call,
)
from mellea.plugins import register

register([
    log_generation_pre_call,
    log_generation_post_call,
])

with mellea.start_session() as m:
    result = m.instruct("...")  # Tracing fires automatically

See docs/docs/how-to/debug-with-plugins.md for complete usage guide and debugging scenarios.
```

### PR DESCRIPTION
**Pull request by akihikokuroda**
# Pull Request

## Issue
Fix: #1250 

## Description

description is in issue #1250 

### Testing
- [ ] Tests added to the respective file if code was changed
- [ ] New code has 100% coverage if code was added
- [ ] Ensure existing tests and github automation passes (a maintainer will kick off the github automation when the rest of the PR is populated)

### Attribution
- [x] AI coding assistants used: claudecode

---

### Adding a new component, requirement, sampling strategy, or tool?

If your PR adds or modifies one of the types below, check the matching box. A checklist of type-specific review items will be posted as a comment.

- [ ] Component
- [ ] Requirement
- [ ] Sampling Strategy
- [ ] Tool

NOTE: Please ensure you have an issue that has been acknowledged by a core contributor and routed you to open a pull request against this repository. Otherwise, please open an issue before continuing with this pull request.

### CODE REVIEW

File: mellea/plugins/builtin_debug/__init__.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+"""Built-in debug plugins for Mellea.
+
+Provides pre-built plugins for common debugging tasks:
+- Generation pipeline tracing (requests, responses, latency, tokens)
+- Sampling strategy diagnostics (iterations, validation, repair, results)
```
Comment (ajbozarth) [inline mellea/plugins/builtin_debug/__init__.py:6]: Module ships three categories — validation is missing here.

```suggestion
Provides pre-built plugins for common debugging tasks:
- Generation pipeline tracing (requests, responses, latency, tokens)
- Validation pipeline tracing (requirements, pass/fail, reasons)
- Sampling strategy diagnostics (iterations, validation, repair, results)
```

File: mellea/plugins/builtin_debug/generation.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+tracking latency, and understanding request/response flow.
+
+Examples:
+    Enable generation tracing:
+
+        from mellea.plugins.builtin_debug import GenerationTracingPlugin
+        from mellea import start_session
+        from mellea.plugins import register
+
+        register([GenerationTracingPlugin()])
```
Comment (ajbozarth) [inline mellea/plugins/builtin_debug/generation.py]: This import fails — `GenerationTracingPlugin` isn't exported from `__init__.py`, and even if it were, the marker class on line 195 has no behavior. Recommend deleting the class (see comment on line 195) and updating this example to the function-based API used everywhere else.

```suggestion
        from mellea.plugins.builtin_debug.generation import (
            log_generation_pre_call,
            log_generation_post_call,
        )
        from mellea import start_session
        from mellea.plugins import register

        register([log_generation_pre_call, log_generation_post_call])
```

File: mellea/plugins/builtin_debug/generation.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    )
+
+
+# Export as a name for convenience
+class GenerationTracingPlugin:
+    """Marker class for GenerationTracingPlugin.
+
+    The actual hooks are registered via @hook decorators above.
+    Use this for reference or type hinting if needed.
+    """
```
Comment (ajbozarth) [inline mellea/plugins/builtin_debug/generation.py]: Marker class with no behavior — passing it to `register()` would be a no-op since hooks are bound by the `@hook` decorator on the functions above. The class is referenced by the module docstring (line 9) and the example file's docstring but never actually used. Recommend deleting; the function-based API is the project's idiomatic pattern.

```suggestion
```

File: mellea/plugins/builtin_debug/generation.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+            pass
+
+    # Fallback to string representation
+    if not text:
+        text = str(action)[:200]
+
+    text = str(text).replace("\n", " ").replace("  ", " ").strip()
+    if len(text) > 100:
+        text = text[:97] + "..."
+    return text
```
Comment (ajbozarth) [inline mellea/plugins/builtin_debug/generation.py:66]: A few cleanups here:

1. `getattr(payload, "action", None)` — `action` is a declared field on `GenerationPreCallPayload`, just use `payload.action`.
2. Reaching into `action._description` is private API; if a description is needed, prefer the public surface or just `str(action)[:100]`.
3. The double-truncation (`[:200]` then `[:97]+'...'`) is unnecessary — pick one.
4. Bare `except Exception` is too wide; narrow to `(AttributeError, TypeError)` or drop it once private-attribute access is removed.

Same comments apply to `_get_response_preview` below.

**Review by ajbozarth — COMMENTED** [review]
Useful addition. Main asks are inline; one cross-cutting item:

**Missing unit tests.** AGENTS.md §10 and the PR template require unit tests for new code; the 7 example scripts are ollama/e2e gated and won't run in the fast tier. Please add `test/plugins/test_builtin_debug.py` with `caplog`-based assertions covering at least:
- Each hook fires and logs the expected line on a synthetic payload
- The empty/missing-field paths the defensive `getattr` code is reaching for (`payload.action = None`, `model_output.value = None`)

Minor: the example block in the PR description has trailing prose inside the code fence — worth fixing so it renders.

File: mellea/plugins/builtin_debug/generation.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        prompt = usage.get("prompt_tokens", "?")
+        completion = usage.get("completion_tokens", "?")
+        return f"({prompt}+{completion}={total})"
+    except Exception:
+        return "unknown"
+
+
+@hook(HookType.GENERATION_PRE_CALL)
+async def log_generation_pre_call(
+    payload: GenerationPreCallPayload, ctx: dict[str, Any]
```
Comment (ajbozarth) [inline mellea/plugins/builtin_debug/generation.py]: `ctx` is a cpex `PluginContext`, not a dict — the body uses attribute access (`ctx.global_context`) which contradicts this annotation. mypy doesn't catch it because the body uses `getattr`/`hasattr`. Same fix needed on `log_generation_post_call` below, and on every hook in `validation.py` and `sampling.py`.

```suggestion
    payload: GenerationPreCallPayload, ctx: Any
```

File: mellea/plugins/builtin_debug/generation.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+            if hasattr(fmt, "args"):
+                requirements = fmt.args.get("requirements", [])
+                repair_text = fmt.args.get("repair", "")
+        except Exception:
+            pass
+
+    # Log main request info
+    prompt_preview = _get_prompt_preview(payload)
+    logger.info(
+        f"[📤 GEN-PRE-CALL gen_id={gen_id}] model={model_id} | prompt={prompt_preview}"
```
Comment (ajbozarth) [inline mellea/plugins/builtin_debug/generation.py]: INFO is too noisy for per-generation events — once registered, every `m.instruct()` call spams two INFO lines. Mellea's `fancy_logger` typically runs at INFO. Recommend DEBUG for routine pre/post lines and reserve INFO for failure summaries (same applies in `validation.py:51,79` and `sampling.py:58,86,89`). The doc on line 277-278 already says "INFO for important events and DEBUG for details" — let's match that.

File: mellea/plugins/builtin_debug/generation.py [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+            logger.debug(f"     {i}. {req_desc}")
+
+    # Log repair feedback if present (indicates a repair attempt)
+    if repair_text:
+        logger.info("   [⭐ REPAIR ATTEMPT] Repair feedback provided:")
+        # Show first 300 chars of repair text
+        repair_preview = repair_text[:300].replace("\n", " ")
+        if len(repair_text) > 300:
+            repair_preview += "..."
+        logger.info(f"   {repair_preview}")
```
Comment (ajbozarth) [inline mellea/plugins/builtin_debug/generation.py]: Repair feedback at 300 chars on INFO can include user PII or upstream model output that landed in the prompt. Recommend DEBUG and/or shortening to ~100 chars to match the prompt-preview cap.

File: mellea/plugins/builtin_debug/generation.py [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+    Args:
+        payload: GenerationPostCallPayload containing model_output, generation_id, latency_ms.
+        ctx: Plugin context for hook execution.
+    """
+    model_output = getattr(payload, "model_output", None)
+    model_id = "unknown"
+    if model_output:
+        gen = getattr(model_output, "generation", None)
+        if gen:
+            model_id = getattr(gen, "model", "unknown")
```
Comment (ajbozarth) [inline mellea/plugins/builtin_debug/generation.py]: Pre-call (line 124) reads `backend_name` from `global_context.state` (which is `backend.model_id`); post-call here reads `model_output.generation.model` (provider-returned, e.g. `ollama_chat/granite4.1:3b`). Logs will look like the model changed mid-call. Pick one source of truth, or label the post-call field differently (e.g. `provider_model=...`).

File: mellea/plugins/builtin_debug/sampling.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+    """Log when repair is triggered during sampling iterations.
+
+    Args:
+        payload: SamplingRepairPayload with repair_iteration, repair_type, failed_validations.
+        ctx: Plugin context for hook execution.
+    """
+    iteration = payload.repair_iteration
+    repair_type = payload.repair_type
+
+    logger.info(f"\n[🔧 REPAIR-TRIGGERED] at iteration {iteration}")
```
Comment (ajbozarth) [inline mellea/plugins/builtin_debug/sampling.py]: Embedded `\n` in log messages breaks log handlers/formatters that expect one record per line (CloudWatch, GCP Logging, structured JSON formatters). Drop the leading `\n` here and on lines 147 and 151 — vertical spacing is the formatter's job, not the message's.

```suggestion
    logger.info(f"[🔧 REPAIR-TRIGGERED] at iteration {iteration}")
```

File: mellea/plugins/builtin_debug/validation.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    ValidationPreCheckPayload,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@hook(HookType.VALIDATION_PRE_CHECK)
+async def log_validation_pre_check(
+    payload: ValidationPreCheckPayload, ctx: dict[str, Any]
+) -> None:
```
Comment (ajbozarth) [inline mellea/plugins/builtin_debug/validation.py:44]: Same `ctx: dict[str, Any]` issue as flagged on `generation.py:112` — applies to every hook function in this file and in `sampling.py`. Will fold all four occurrences into one fix.

File: docs/docs/how-to/debug-with-plugins.md [hunk 11]
```diff
@@ -0,0 +1,309 @@
+---
+title: "Debug with Plugins"
+description: "Use built-in debug plugins to trace generation, validation, and sampling behavior in detail."
+# diataxis: how-to
+---
+
+**Prerequisites:** [The Requirements System](../concepts/requirements-system), `pip install mellea`.
```
Comment (ajbozarth) [inline docs/docs/how-to/debug-with-plugins.md]: AGENTS.md §12 requires `.md` extensions on internal cross-doc links — without it, the broken-link check fails when the site is built with a non-root baseUrl.

```suggestion
**Prerequisites:** [The Requirements System](../concepts/requirements-system.md), `pip install mellea`.
```

File: docs/docs/how-to/debug-with-plugins.md [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+- No plugins fire in the hot path when not registered
+
+For production use, you can safely leave plugins registered — they only log when
+enabled. For maximum performance, simply don't register them.
+
+## Next steps
+
+- [Observability: Tracing](../observability/tracing.md) — export traces to Jaeger or Grafana
+- [Handling Exceptions and Failures](./handling-exceptions.md) — work with sampling failures
+- [The Requirements System](../concepts/requirements-system) — understand validation in depth
```
Comment (ajbozarth) [inline docs/docs/how-to/debug-with-plugins.md]: Same `.md` rule as line 7.

```suggestion
- [The Requirements System](../concepts/requirements-system.md) — understand validation in depth
```

File: mellea/plugins/builtin_debug/generation.py (hunk 2 above)
Comment (planetf1) [inline mellea/plugins/builtin_debug/generation.py]: Heads up — this example will fail for anyone who copies it. `GenerationTracingPlugin` isn't exported from `mellea.plugins.builtin_debug` (it's not in `__init__.py`), so the import raises `ImportError`. Even importing directly from `.generation`, the class has no `@hook` methods so `register()` raises `TypeError`.

The example files in `docs/examples/plugins/` all do this correctly — aligning the docstring would make it consistent:

```python
from mellea.plugins.builtin_debug.generation import (
    log_generation_pre_call,
    log_generation_post_call,
)
from mellea.plugins import register

register([log_generation_pre_call, log_generation_post_call])
```

The `GenerationTracingPlugin` class at the bottom of the file could probably just be removed — it has no behaviour and its presence implies an API that doesn't hold up.

File: mellea/plugins/builtin_debug/generation.py (hunk 5 above)
Comment (planetf1) [inline mellea/plugins/builtin_debug/generation.py]: `ctx` here is actually a `PluginContext` object — the code a few lines down accesses `ctx.global_context.state` (attribute access), which works at runtime but contradicts the `dict[str, Any]` annotation. Someone writing their own hook by copying this signature would reach for `ctx["something"]` and get a `TypeError`.

The project convention in `base.py` and the test suite uses `ctx: Any` for standalone hook functions — same applies to the other seven signatures across `sampling.py` and `validation.py`.

File: docs/docs/how-to/debug-with-plugins.md (hunk 11 above)
Comment (planetf1) [inline docs/docs/how-to/debug-with-plugins.md]: Small one — the project uses `uv` rather than `pip`, and `register()` also needs the `hooks` extra or it raises `ImportError`. Suggest:

```
uv add 'mellea[hooks]'
```

Also, the two `../concepts/requirements-system` links (here and line 309) are missing `.md` — they'll fail the broken-link check on the fork deploy. The other links on line 309 already have the extension, so just needs `../concepts/requirements-system.md` in both places.

File: docs/examples/plugins/builtin_generation_tracing.py [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+from mellea.plugins.builtin_debug.generation import (
+    log_generation_post_call,
+    log_generation_pre_call,
+)
+
+logging.basicConfig(level=logging.INFO, format="%(levelname)s: %(message)s")
+log = logging.getLogger(__name__)
+
+# Enable generation tracing by registering the hooks
+register([log_generation_pre_call, log_generation_post_call])
```
Comment (planetf1) [inline docs/examples/plugins/builtin_generation_tracing.py]: `register()` at module level is fine for the examples here since pytest runs them in separate processes. For users copying this pattern into a notebook or long-running app though, hooks would be installed globally with no cleanup. `plugin_scope` (already in `mellea.plugins`) handles this cleanly — some examples in this directory already use it:

```python
from mellea.plugins import plugin_scope

def main():
    with plugin_scope([log_generation_pre_call, log_generation_post_call]):
        with mellea.start_session() as m:
            ...
```

File: mellea/plugins/builtin_debug/generation.py [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+    # Try format_for_llm() for structured output
+    if not text and hasattr(action, "format_for_llm"):
+        try:
+            formatted = action.format_for_llm()
+            # Extract args if available
+            if hasattr(formatted, "args"):
+                desc = formatted.args.get("description", "")
+                if desc:
+                    text = str(desc)[:200]
+        except Exception:
```
Comment (planetf1) [inline mellea/plugins/builtin_debug/generation.py]: Silent `except` is the right call here (debug plugins shouldn't blow up a generation call), but `pass` also swallows bugs in the plugin itself. Swapping for a `logger.debug` would make the plugin self-diagnosable with basically no extra cost — same pattern applies to the other three `except` blocks in this file (lines 83, 106, 138):

```python
except Exception:
    logger.debug("preview extraction failed", exc_info=True)
```

File: mellea/plugins/builtin_debug/__init__.py [hunk 15]
```diff
@@ -0,0 +1,55 @@
+"""Built-in debug plugins for Mellea.
+
+Provides pre-built plugins for common debugging tasks:
+- Generation pipeline tracing (requests, responses, latency, tokens)
```
Comment (planetf1) [inline mellea/plugins/builtin_debug/__init__.py:4]: Minor: the docstring only mentions generation and sampling, but the module also exports the validation hooks. Worth adding a bullet for completeness.

**Review by planetf1 — CHANGES_REQUESTED** [review]
Two things worth flagging at the PR level:

**Missing sidebar entry:** `docs/docs/how-to/debug-with-plugins.md` isn't listed in `docs/sidebars.ts`, so it'll build fine but won't appear in the site navigation. Needs an entry like `'how-to/debug-with-plugins'` added to the how-to array.

**No test coverage:** The new `builtin_debug` modules don't have any tests — `test/plugins/` would be a natural home. The three helper functions (`_get_prompt_preview`, `_get_response_preview`, `_get_token_usage`) have branchy logic that's easy to unit-test with synthetic payloads, and a quick import smoke test would have caught the `GenerationTracingPlugin` issue below. Worth a `test/plugins/test_builtin_debug.py`.

File: test/plugins/test_builtin_debug.py [hunk 16]
```diff
@@ -0,0 +1,600 @@
...
+    log_validation_post_check,
+    log_validation_pre_check,
+)
+from mellea.plugins.builtin_debug.generation import (
+    _get_prompt_preview,
+    _get_response_preview,
+    _get_token_usage,
+)
+from mellea.plugins.builtin_debug.sampling import log_sampling_iteration
+from mellea.plugins.builtin_debug.validation import log_validation_post_check
```
Comment (ajbozarth) [inline test/plugins/test_builtin_debug.py]: Already imported from `mellea.plugins.builtin_debug` at lines 23 and 27 — these two lines can be dropped.

```suggestion
```

**Review by ajbozarth — COMMENTED** [review]
Some feedback from Claude.

Prior review items addressed; tests pass.

`plugin_scope` was applied to `builtin_generation_tracing.py` but not the other five
`builtin_*.py` examples — they still call `register()` at module scope, which leaks
hooks if anyone copies them into a notebook or long-running app. Worth converting all
five to match the generation example, e.g.:

    # at top
    from mellea.plugins import plugin_scope

    # in main()
    with plugin_scope([log_validation_pre_check, log_validation_post_check]):
        with mellea.start_session() as m:
            ...

Files: builtin_complete_diagnostics.py, builtin_full_pipeline_tracing.py,
builtin_sampling_diagnostics.py, builtin_validation_failures.py,
builtin_validation_strict.py, builtin_validation_tracing.py.

**Review by ajbozarth — APPROVED** [review]

**Comment by planetf1** [comment]
Hi @akihikokuroda — one thing still open: the `generation.py` module docstring example imports and registers `GenerationTracingPlugin`, but it's not exported from `__init__.py` (ImportError), and even if it were the class is a no-op marker — hooks are wired via `@hook` decorators, so `register([GenerationTracingPlugin()])` doesn't enable tracing. The correct pattern is already in `__init__.py`'s docstring; the `generation.py` example just needs to match it.

(`ctx: dict[str, Any]` is also wrong across the three modules — it's a `PluginContext` — happy to take that as a follow-up.)

**Comment by akihikokuroda** [comment]
@planetf1 Thanks for review.  The first issue has been addressed.  Please check again.  I'm updating the second issue.

**Review by planetf1 — APPROVED** [review]
All points from my earlier CHANGES_REQUESTED review have been addressed in the follow-up commits — thank you for the thorough response:

✅ `GenerationTracingPlugin` broken import → replaced with function-based API  
✅ `ctx: dict[str, Any]` → `ctx: PluginContext` via `TYPE_CHECKING` in all three modules  
✅ Sidebar entry added  
✅ Unit test file added (`test/plugins/test_builtin_debug.py`, 598 lines)  
✅ `pip install mellea` → `uv add 'mellea[hooks]'` in docs prerequisites  
✅ Missing `.md` extension on Requirements System cross-links → fixed  
✅ Bare `register()` at module scope → `plugin_scope` context manager in all six example files  
✅ Validation pipeline bullet added to `__init__.py` module docstring  
✅ Silent `except` → `logger.debug(..., exc_info=True)`  
✅ Leading `\n` in `sampling.py` log messages → removed  
✅ Log levels: routine output demoted to DEBUG; key events remain at INFO  
✅ PII risk on repair text: capped to 100 chars at DEBUG level  

Two minor nits for a follow-up (not blocking):

- `from typing import Any` in `test_builtin_debug.py` — appears unused; ruff should flag it
- `hook` in `from mellea.plugins import HookType, hook, register` — also appears unused in the test body

Happy to merge.

File: test/plugins/test_builtin_debug.py [hunk 17]
```diff
@@ -0,0 +1,598 @@
+"""Tests for built-in debug plugins (generation, sampling, validation).
+
+Verifies that each built-in debug plugin hook fires correctly and logs expected
+output, including edge cases with missing or None-valued payload fields.
+"""
+
+from __future__ import annotations
+
+import logging
+from typing import Any
```
Comment (planetf1) [inline test/plugins/test_builtin_debug.py]: NIT: `Any` is not referenced anywhere in the test body — unused import.

```suggestion
```

File: test/plugins/test_builtin_debug.py [hunk 18]
```diff
@@ -0,0 +1,598 @@
...
+from unittest.mock import MagicMock
+
+import pytest
+
+pytestmark = pytest.mark.unit
+
+pytest.importorskip("cpex.framework")
+
+from mellea.plugins import HookType, hook, register
```
Comment (planetf1) [inline test/plugins/test_builtin_debug.py]: NIT: `hook` is imported but never used in any test function — only `HookType`, `register`, and `invoke_hook` are referenced.

```suggestion
from mellea.plugins import HookType, register
```
